enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1156 @@
+1. If persisted version is on a list of blocked version migrations, **refuse to start MicroShift**.
+1. If binary is the same version as persisted in metadata, **no need for a data migration**.
+1. Otherwise attempt to migrate the data.
+
+### Open Questions [optional]
+
+#### Should 4.13 -> 4.14 be supported?
+- Even if we handle "no metadata, no backup, existing data, no next-boot-action" and make
+  a backup, upon rollback 4.13 won't be able to restore it, so it would try to use 4.14's
+  data.
+  - Is manual intervention acceptable (admin manually copying backup to data)?
+  - We most likely don't want to implement special case for that (`pre-run` would save
+    that persist it was 4.13 previously and red script would restore before shutting down)
+
+#### Should MicroShift healthcheck check and log version skew problems so it's easier to debug?
+- Why not?
+
+### Workflows in detail
+
+#### Decision tree
+
+Metadata schema reminder
+```json
+// /var/lib/microshift-backups/health.json
+{
+  "commits": [
+    {
+      "id": "rhel-d1",
+      "microshift": "unknown | healthy | unhealthy",
+      "system": "unknown | healthy | unhealthy",
+      "last_boot": "yyyy-mm-dd HH:MM:SS"
+    },
+    {
+      "id": "rhel-d0",
+      "microshift": "unknown | healthy | unhealthy",
+      "system": "unknown | healthy | unhealthy",
+      "last_boot": "yyyy-mm-dd HH:MM:SS"
+    }
+  ]
+}
+
+// /var/lib/microshift/version.json
+{
+  "microshift": "4.14.0",
+  "commit": "rhel-d1"
+}
+```
+
+Abbreviations, glossary:
+- `microshift/` -> `/var/lib/microshift/`
+- `version.json` -> `/var/lib/microshift/version.json`
+- `backups/` -> `/var/lib/microshift-backups/`
+- `health.json` -> `/var/lib/microshift-backups/health.json`
+- `prev-boot-commit` - commit in `health.json` with most recent `boot` timestamp in `health.json`
+- `current-commit` - currently running commit
+  - if absent in `health.json`, it means it's 1st boot of the commit
+  - if present, it's a subsequent boot and will be the same as `prev-boot-commit`
+- `earlier-commit` - commit that was before `prev-boot-commit`
+  - if `current-commit` exists in `health.json`, then this is commit that system is being
+    upgraded from and will automatically roll back to
+  - if `current-commit` is not present in `health.json` (so ostree history looks like `current`,
+    `prev-boot`, and `earlier` commits), it might no longer exist according to ostree
+    (only `current` and `prev-boot` which is a rollback commit)
+
+`microshift pre-run`:
+1. Neither `microshift/` nor `backups/` exist
+   > *First boot*
+   - **exit 0** :leftwards_arrow_with_hook:
+
+1. `microshift/` exists and `version.json` does not:
+   > *Assume it's 4.13*
+   - Backup data to `backups/4.13/`
+   - Proceed to data migration
+
+1. Load `version` and `health.json`
+
+1. `prev-boot-commit.system == healthy`
+   - Backup to `microshift-backups/prev-boot-commit.id`
+   - Special case - rollback on demand
```

Comment:
> Does this special case actually come after the backup step on line 458? Or should it be evaluated first?

If prev-boot was healthy, then we want to make a backup first and then restore `current`.
This will look slightly different in simplified version